src/ipc/ipc3/dai.c

```diff
@@ -58,6 +58,18 @@ int dai_config_dma_channel(struct comp_dev *dev, void *spec_config)
 		handshake = dai_get_handshake(dd->dai, dai->direction,
 					      dd->stream_id);
 		channel = EDMA_HS_GET_CHAN(handshake);
+#if CONFIG_IMX8ULP
+		switch (dai->dai_index) {
```

Comment:
I still think we can do better. dai.c is generic code so it should not contain any specific ifdefs.

What do you think about

```
--- a/src/drivers/imx/edma.c
+++ b/src/drivers/imx/edma.c
@@ -287,6 +287,8 @@ static int edma_setup_tcd(struct dma_chan_data *channel, int16_t soff,
 {
 	int rc;
 	uint32_t sbase, dbase, total_size, elem_count, elem_size, size;
+	struct dai_data *dd = channel->dev_data;
+	int direction, handshake, dmamux;
 
 	assert(!sg);
 	assert(cyclic);
@@ -332,14 +334,22 @@ static int edma_setup_tcd(struct dma_chan_data *channel, int16_t soff,
 	if (rc < 0)
 		return rc;
 
+
 #ifdef CONFIG_IMX8ULP
+
+	if (channel->direction == DMA_DIR_MEM_TO_DEV)
+		direction = DAI_DIR_PLAYBACK;
+	else
+		direction = DAI_DIR_CAPTURE;
+
+	handshake = dai_get_handshake(dd->dai, direction, 0);
+	dmamux = EDMA_HS_GET_DMAMUX_CFG(handshake);
+
 	/* Do not write EDMA_CH_MUX register when it has value,
 	 * otherwise the register will be cleared.
 	 */
 	if (!dma_chan_reg_read(channel, EDMA_CH_MUX))
-		(channel->direction == DMA_DIR_MEM_TO_DEV) ?
-			dma_chan_reg_write(channel, EDMA_CH_MUX, IMX8ULP_DMAMUX2_SAI5_TX) :
-			dma_chan_reg_write(channel, EDMA_CH_MUX, IMX8ULP_DMAMUX2_SAI5_RX);
+			dma_chan_reg_write(channel, EDMA_CH_MUX, dmamux);
 #endif
 
 	chan_addr_convert(channel, &sbase, &dbase);
diff --git a/src/include/sof/drivers/edma.h b/src/include/sof/drivers/edma.h
index 5019923f9..16a069500 100644
--- a/src/include/sof/drivers/edma.h
+++ b/src/include/sof/drivers/edma.h
@@ -82,6 +82,9 @@
 #define EDMA_HS_SET_IRQ(irq) SET_BITS(8, 0, irq)
 #define EDMA_HS_GET_CHAN(hs) (((hs) & MASK(13, 9)) >> 9)
 #define EDMA_HS_SET_CHAN(chan) SET_BITS(13, 9, chan)
+#define EDMA_HS_GET_DMAMUX_CFG(hs) (((hs) & MASK(21, 14)) >> 14)
+#define EDMA_HS_SET_DMAMUX_CFG(cfg) (chan) SET_BITS(21, 14, cfg)
+
 #define EDMA_HANDSHAKE(irq, channel)\
 	(EDMA_HS_SET_CHAN(channel) | EDMA_HS_SET_IRQ(irq))
 
diff --git a/src/platform/imx8ulp/lib/dai.c b/src/platform/imx8ulp/lib/dai.c
index 282201b16..29b16cc86 100644
--- a/src/platform/imx8ulp/lib/dai.c
+++ b/src/platform/imx8ulp/lib/dai.c
@@ -29,14 +29,14 @@ static SHARED_DATA struct dai sai[] = {
 			.depth		= 16,  /* in 4 bytes words */
 			.watermark      = 8,
 			.handshake	= EDMA_HANDSHAKE(IMX8ULP_EDMA2_CHAN0_IRQ,
-							 IMX8ULP_EDMA2_CHAN0),
+							 IMX8ULP_EDMA2_CHAN0, IMX8ULP_DMAMUX2_SAI5_TX),
 		},
 		.fifo[SOF_IPC_STREAM_CAPTURE] = {
 			.offset		= SAI_5_BASE + REG_SAI_RDR3,
 			.depth		= 16,  /* in 4 bytes words */
 			.watermark      = 8,
 			.handshake	= EDMA_HANDSHAKE(IMX8ULP_EDMA2_CHAN1_IRQ,
-							 IMX8ULP_EDMA2_CHAN1),
+							 IMX8ULP_EDMA2_CHAN1, IMX8ULP_DMAMUX2_SAI5_RX),

```

This is not tested. Just for you to make an idea of the direction I want to set for the change.